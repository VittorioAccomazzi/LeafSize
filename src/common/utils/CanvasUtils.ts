import {Canvas} from 'canvas'
import ColourModels from '../imgLib/ColourModels';
import { ImageUint8 } from '../imgLib/ImageBase';
import { isCanvas } from './TestUtils';

export default class CanvasUtils {
    private constructor() {} // singleton

    static GetImageData ( inCanvas : Canvas | HTMLCanvasElement ) : ImageData {
        const width = inCanvas.width;
        const height= inCanvas.height;
        let data : ImageData;
        if( isCanvas( inCanvas ) ){ // this if is necessary becaue the functon below are NOT equivalent for type script.
            const ctx = inCanvas.getContext('2d');
            data = ctx.getImageData(0,0, width, height);
        } else {
            const ctx = inCanvas.getContext('2d');
            data = ctx!.getImageData(0,0, width, height); 
        }
        return data
    }

    static PutImageData ( inCanvas : Canvas | HTMLCanvasElement, data : ImageData  ) : void {
        inCanvas.width = data.width;
        inCanvas.height= data.height;
        if( isCanvas( inCanvas ) ){ // this if is necessary becaue the functon below are NOT equivalent for type script.
            const ctx = inCanvas.getContext('2d');
            ctx.putImageData(data,0,0)
        } else {
            const ctx = inCanvas.getContext('2d');
            ctx!.putImageData(data,0,0)
        }
    }
    
    /**
     * converts an HTML canvas object to a  node canvas object
     * @param image 
     * @returns 
     */
    static fromHTML( image : HTMLCanvasElement ) : Canvas {
        const width = image.width;
        const height= image.height;
        const canvas= new Canvas(width, height);
        const data  = this.GetImageData(image);
        this.PutImageData(canvas, data);
        return canvas;
    }
    
    /**
     * converts a node canvas object to a HTML canvas object
     * @param image 
     * @returns 
     */
    static toHTML( image : Canvas ) : HTMLCanvasElement {
        const width = image.width;
        const height= image.height;
        const htmImg= document.createElement('canvas');
        htmImg.width = width;
        htmImg.height= height;
        const data   = this.GetImageData(image);
        this.PutImageData(htmImg, data); 
        return htmImg; 
    }

    /**
     * converts a gray scale 8 bit image to a canvas object
     * @param img8 
     * @returns 
     */
    static toCanvas( img8 : ImageUint8 ) : Canvas {
        const canvas = new Canvas(img8.width, img8.height);
        const imgData= ColourModels.fromRGB(img8, img8, img8);
        this.PutImageData(canvas, imgData);
        return canvas;
    }
}