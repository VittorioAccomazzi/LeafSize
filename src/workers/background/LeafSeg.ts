import ColourModels from "../../common/imgLib/ColourModels";
import Mask from "../../common/imgLib/Mask";
import { Bbox, Colour } from "../../common/imgLib/Types";

export enum BackgroundType {
    Image,
    Transparent
}

/**
 * This class perform the segmentation of the leaf, given the image and thresholds
 */
export default class LeafSeg {

    private constructor( ) { } // singleton

    /**
     * Perform the identification of the leaf(s) in the input image
     * @param imgData  input image
     * @param hueThr  hue threshold [0,255]
     * @param satThr saturation threshold [0,255]
     * @param nLeafs number of leafs 1 or 2
     * @returns areas : array with the areas measured, bboxs : bounding box for each leaf measured.
     */
    static Process(imgData : ImageData, hueThr : number, satThr : number, nLeafs : number, backgroundType: BackgroundType = BackgroundType.Image ) :  { areas : number[], bboxs : Bbox[] }  {
        let areas : number [] = [];
        let bboxs : Bbox[]    = [];

        const {hue,sat} = ColourModels.toHSV(imgData);
        const rMask = new Mask(imgData.width, imgData.height);
        const mPixels = rMask.imagePixels;
        const hPixels = hue.imagePixels;
        const sPixels = sat.imagePixels;

        // background mask
        hueThr *= 360/255; // rage expected by the library
        satThr *= 1/255;
        hPixels.forEach((v,i)=>mPixels[i]= v > hueThr || sPixels[i] < satThr); 

        // fill holes
        const bMask = rMask.Fill(1,1);
        const dMask = bMask;

        // foregound mask
        dMask.Not();

        // extract components
        const leafs = dMask.Components(nLeafs);

        // overlay
        const colours : Colour [] = [{r:255,g:0,b:0},{r:255,g:255,b:0}];

        // set background
        LeafSeg.setBackground(backgroundType, imgData);

        leafs.forEach((m, i)=>{
            const col  = colours[i%colours.length];
            const area = m.area;
            const bbox = m.Boundaries();
            areas.push(area);
            bboxs.push(bbox);
            const dil  = m.Dilate(1);
            dil.Minus(m); // this is the outline.
            dil.Overlay(imgData, col);  
        })

        return { areas, bboxs};
    }


    private static setBackground(backgroundType: BackgroundType, imgData: ImageData) {
        const len = backgroundType === BackgroundType.Image ? 0 : imgData.data.length;
        let data= imgData.data;
        let ptr = 0;
        while (ptr < len) {
            data[ptr++] = 0;
            data[ptr++] = 0
            data[ptr++] = 0;
            data[ptr++] = 0;
        }
    }
}