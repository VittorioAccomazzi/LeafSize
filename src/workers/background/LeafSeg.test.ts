import hash, { loadCanvas, testImage } from '../../common/utils/TestUtils';
import ImageLoader, { ImgLoaderFile, ImgLoaderFileHandle } from '../foreground/ImageLoader';
import CanvasUtils from '../../common/utils/CanvasUtils';
import LeafSeg, { BackgroundType } from './LeafSeg'


describe('LeafSeg',()=>{

    class EmptyFileMock implements ImgLoaderFile  {
        constructor() {}
        async arrayBuffer() {
            return new ArrayBuffer(0)
        }
    }
    
    class FileHandleMock implements ImgLoaderFileHandle {
        public name : string = 'image';
        async getFile () { return new EmptyFileMock() }
    }

    const getImageFromFile = async ( ) : Promise<ImageBitmap> => {
        const image = await loadCanvas(testImage);
        return image as unknown as ImageBitmap; 
    }
    const targetSize = 1024
    const files = [new FileHandleMock() ]
    const emptySet = new Set<number>();

    test('shall measure the leaf correctly', async()=>{

        const imgLoader = new ImageLoader(files, 4, targetSize, getImageFromFile);

        expect(imgLoader.List.length).toBe(4);

        const htmlCnv = document.createElement('canvas');
        const hueThr = 68;
        const satThr = 91;
        const refArea= [[421701,365059], [514353,351194], [246855,197252], [472529,387532]];
        const error = 2; // in persentage

        for( const [idx, name] of imgLoader.List.entries() ){
            const { scale, imgData } = await imgLoader.getImage(idx);
            const {areas} = await LeafSeg.Process(imgData!, hueThr, satThr, 2, emptySet, emptySet);
            CanvasUtils.PutImageData(htmlCnv,imgData!);
            const hsh     = await hash(htmlCnv, name );
            areas.forEach((area,index)=>{
                const rescaled = area / ( scale * scale );
                const target   = refArea[idx][index];
                const percErr  = Math.abs(rescaled-target)/target *100;
                expect(percErr).toBeLessThan(error);
            })
            expect(hsh).toMatchSnapshot();
        }
    })

    test('shall retun a transparent image', async ()=>{
        const imgLoader = new ImageLoader(files, 1, targetSize, getImageFromFile);

        expect(imgLoader.List.length).toBe(1);

        const htmlCnv = document.createElement('canvas');
        const hueThr = 68;
        const satThr = 91;
        const { imgData } = await imgLoader.getImage(0);
        await LeafSeg.Process(imgData!, hueThr, satThr, 2, emptySet, emptySet, BackgroundType.Transparent);
        CanvasUtils.PutImageData(htmlCnv,imgData!);
        const hsh     = await hash(htmlCnv, 'transparent' );
        expect(hsh).toMatchSnapshot();
    })

})