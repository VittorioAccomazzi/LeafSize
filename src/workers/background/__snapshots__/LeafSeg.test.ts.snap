// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeafSeg shall measure the leaf correctly 1`] = `"ffffffffffffffffffd4000000000000ffffffffffffffffffc0000000000000fffffffffffffffffe40000000000000ffffffffffffffffff00000000000000fffffffffffffffff800000000000000ffffffffffffffffc000000000000000fffffffffffffff00000000000000000ffffffffffffffc00000000000000000ffffffffffffff800000000000000000fffffffffffffc000000000000000000fffffffffffff0000000000000000000ffffffffffffc4004000000000000000ffffffffffff901fff00000000000000fffffffffffe60ffffc0000000000000fffffffffffc83fffff4000000000000fffffffffffb07fffffc000000000000ffffffffffe61fffffff000000000000ffffffffffcc3fffffffc00000000000ffffffffffd8ffffffffc00000000000ffffffffffb0ffffffff800000000000ffffffffff61fffffff9800000000000fffffffffee3fffffff8000000000000fffffffffdc7ffffff50000000000000fffffffffd87ffffff00000000000000fffffffffb07fffffe00000000000000fffffffff70f7ffff600000000000000fffffffff61f7abff000000000000000ffffffffec3e001ff000000000000000ffffffffdc7e000fb000000000000000ffffffffd97c0007b000000000000000fffffffff8fc00018000000000000000ffffffffb3fe00008000000000000000ffffffffffff0003fffe00000c3ffff4fffffffffffe0001fffe0000043fffb2fffffffffffe0000fffe0000061ffdfcfffffffffffc00007ffc0000021ffff8fffffffffffc00003ffc0000021fffe1fffffffffff800003fd00000000ffff0fffffffffff000003fc00000000fffccfffffffffff000003fc000000007fff0ffffffffffe000007f8000000007fff2ffffffffffe000007f8000000007fff4ffffffffffc000007f0000000047fff4ffffffffffc000007e0000000043ffe2ffffffffff8000007e0000000063ffc2ffffffffff800000200000000063ffe7ffffffffffc00000200000000063ffc6ffffffffff800000002000000071ffecffffffffff800000007800000031ff98ffffffffff000000007800000031ff1cfffffffffe00000000f800000031fe9dfffffffffc00000000f800000030ffbefffffffffc00000000f800000010fffefffffffff800000000f000000010fff8fffffffffc40000001f000000010fffeffffffffff80000001e000000010fffaffffffffff80000001c000000010ffffffffffffff80000003c0000000107ff4ffffffffff00000007c0000000007fdaffffffffff00000007c0000000087fedffffffffff0000000fc0000002187fc8fffffffffe0000000f80000002987ff8fffffffffe0000000f80000006887f38fffffffffc0000001f000000069c7fd8fffffffffc0000001f00000006983fd0fffffffffc0000001f00000006983fe2fffffffffc0000001f00000006d87fa8ffffffffff0000001e00000006887ffcffffffffff0000001e00000006887ff8ffffffffff0000001e00000004887fb1ffffffffff0000001f80000004887f29ffffffffff0000001fe0000004987fd4ffffffffff8000001fe0000004987ff4ffffffffff8000003ff0000004987ff1ffffffffff8000007ff0000004983ef0ffffffffff800000fff8000000907fd8ffffffffff800001fff8000000905f88ffffffffff800003fff0000000107ec4ffffffffff800003fff0000000107e40ffffffffff800007fff80000001076c0ffffffffff800007fff8000001107fc0ffffffffff800007fff8000001107e40ffffffffff800007fffc000000005c00ffffffffff80000ffffc00000000fec0ffffffffff80000ffff800000000f1c0ffffffffff80000ffff8000000206500ffffffffff80001ffff8000000002460ffffffffff80001ffff8000000003640ffffffffffc0001ffff4000000403ec4ffffffffffc0001ffff4000000005464ffffffffffc0003ffff2000000003308ffffffffffc0003ffff2000000010080ffffffffffc0007ffff8000000019b80ffffffffffc0007ffff800000003da80ffffffffffc0007ffff0000000026a00ffffffffffe000fffff0000000009e00ffffffffffe0007fff40000000000000ffffffffffe0007fff40000000000000ffffffffffe0007fff80000000000000ffffffffffe000ffffa0000000000000ffffffffffe000ffff80000000000000ffffffffffe000ffff90000000000000ffffffffffc080ffff80000000000000ffffffffffe0c1ffffc0000000000000fffffffffff1c1ffffe0000000000000fffffffffff1c1fffff0000000000000ffffffffffffc1fffff0000000000000ffffffffffffe1fffff0000000000000ffffffffff7ff1fffff0000000000000ffffffffffbf1bfffa30000000000000ffffffffffd20bfffd10080000000000ffffffffffc00bffff08080000000000ffffffffffe009fffe80000000000000fffffffffff001fffe84400000000000fffffffffff800fffb2e800000000000fffffffffffc00bfcb0e000000000000ffffffffffff003dd104000000000000ffffffffffff80004000000000000000ffffffffffffc0000000000000000000fffffffffffff0000000000000000000fffffffffffffc000000000000000000ffffffffffffff800000000000000000fffffffffffffff80000000000000000fffffffffffffffffc00000000000000fffffffffffffffffffe000000000000ffffffffffffffffffff000000000000ffffffffffffffffffff000000000000ffffffffffffffffffff000000000000"`;

exports[`LeafSeg shall measure the leaf correctly 2`] = `"fffffd4000003ffffe00bfffffff004ffffe80800000801c00007fffffff004ffffef0000000000000007fffffff004fffffb1000040000000001fffffff004fffffff800000000000001fffffff004ffffff60000000000000003ffffff00cfffffde0100004000000001ffffff00cffffffa00000000000000007fffff00dffffffe100004000000001f3fffff00dffffffd00000c0003f8003fffffff00dffffffc030030000fff001fffffff00dfffffff0300400047fe000fffffff005fffffffdb00800010060007feffff005fffffeffa038000400f8003ffffff005fffffffb402000002321001ffbfff005ffffffffc0c000008006c00ff9fff005ffffffff01c000060049b007fffff03fffffffff0380000001c0a003fffff07ffffffffe070001000305f801fffff06ffffffffe0f0000000e4ff801fffff07ffffffffc1e00000090f0ff00fffff07ffffffffc1c0000000361ff007ffff07ffffffff8380000000040ff007ffff07ffffffff07800080000c0bf803ffff0cffffffff0701000000000ffe01ffff087fffffff0e020180000002fe05ffff00ffffffff1e0000000000003e00ffff00fffffffe1c0000000000001f82ffff0cfffffffc1c0000000000001f007fff0ffffffffc1c0800000000000f817fff0ffffffff8381000000000001f007fff0ffffffff8381000000000000fc03fff0ffffffffdf83000000000001ff1bfff0ffffffffff82000000000001ff9dfff0ffffffffff06000000000000ff8dfff0ffffffffdf04000000000000ffcffff0ffffffffff0c0000000000007fcffff0cffffffffe0c0000000000007feefff0cffffffffe080000000000003feffff00fffffffbc180000000000007ff7fff0cffffffffc100000000000003ff7fff1dffffffefc100000000000003ff7fff1fffffffffc700000000000003ffffff1ffffffff7c700000000000003ffffff1ffffffff7ce00000000000003ffffff1fffffffff8e00000000000001ffffff0ffffffff7ce00000000000001ffffff1fffffffefcc00000000000001ffffff1fffffffef8c00000000000001ffffff1dffffffef9c00000000000001ffffff1cffffffef9c00000000000001ffffff1fffffffef9c00000000000003ffffff00ffffffef9c00000000000001ffffff1dffffffff1800000000000003ffffff1fffffffff1800000000000001ffffff1fffffffff1800000000000001ffffff1fffffffff1800000000000001ffffff1fffffffff3800000000000000ffffff0fffffffff1800000000000000ffffff1fffffffff3800000000000000ffffff1fffffffff18000000000000007ffffd1effffffff08000000000000007ffff9c0ffffffff08000000000000003ffff9ffffffffff08000000000000003ffff9ffffffffff8c000000000000003ffff9ffffffffff8c000000000000003ffff9ffffffffff8c000000000000007ffff9ffffffffff0600000000000000fffff9ffffffffef0600000000000001fffff9f9ffffffef0600000000000001fffff8f8ffffffef1700000000000001fffffcc0ffffffef0700000000000001fffffce9f1ffffef0700000000000001fffffcf3e0ffffff0700000000000001fffffce0c8ffffff0700000000000001fffffce0e8ffffff0300000000000003fffffce0e8fffff70380000000000003fffffcf1e8fffff703c0000000000003fffffcffe1fffff70300000000000003fffffcfff1fffff30380000000000003fffffcfffffffffb8380000000000003fffffcfffffffffb8380000000000003fffffcffffffffeb81c0000000000003fffffc7ffffffffb8180000000000003fffffc7ffffffffd81c0000000000003fffffc7ffffffff980c0000000000003fffffd7ffffffff1c0e0000000000003fffffd7fffffffffc080000000000003fffffc7fffffffffc000000000000003fffffe7ffffffffe0000000000200003ffffff7fffffffff0000000000000007ffffff1fffffffff0000000000300007ffffff1fffffffffa000000000f00007ffffff1fffffffffa000000000f80007ffffff1fffffffff9000000007fc0007ffffff0fffffffffd000000000fc000fffffff1fffa000000000000000080003caffff1fff00000000000000007c00031dffff1eff000000000000000014000315ffff1cff40000000000000002e00032bffff1dffc000000000000000fe000627ffff00ffc000000000000000df000c17ffff1dff08000000000000001e00000fffff1fff48000000008000007b80000fffff1fff00000000000000001fc0001fffff1fffc0000000000000007bd8013fffff1fff00000000000000003ff8017fffff0fff000000020010000017f8007fffff1fff000000000000000006d900ffffff1fff000000020000000010e001ffffff1fff0000000000000000000003ffffff1cff0000000000000000000007ffffff1cffd80000041c00000000000fffffff00ff88000001dc00000000003fffffff0cff920000017700000000007fffffff0fff8000000427c000000000ffffffff1fff0040000004e000000003ffffffff1ffffff796000c700000000fffffffff0ffffafff63ffffe0000003fffffffff0fffbffef79bbffffff003ffffffffff0ffffbeb36d83fffffffffffffffffff0ffff3fff71a7fffffffffffffffffff1ffffd5ce71bffffffffffffffffffff1cffff47e2ffbfffffffffffffffffff0cfffbe5c6edafffffffffffffffffff00ffffec20597fffffffffffffffffff0cfff7ff0000ffffffffffffffffffff0fffffacc00003ffffffffffffffffff0fff"`;

exports[`LeafSeg shall measure the leaf correctly 3`] = `"ffffffffffffffffffffc00000000000ffffffffffffffffffffc00000000000ffffffffffffffffffffe00000000000ffffffffffffffffffffc80000000000ffffffffffffffffffffc00000000000ffffffffffffffffffffd00000000000ffffffffffffffffffffe00000000000ffffffffffffffffffffe00000000000fffffffffffffffffffff00000000000ffffffffffffff0003fff80000000000fffffffffffff8000000f80000000000ffffffffffffc00000000c0000000000ffffffffffff00000000060000000000fffffffffffe00000000000000000000fffffffffff800400000000000000000fffffffffff006208000000000000000ffffffffffe00e620000000000000000ffffffffffc19eca0000000000000000ffffffffff83d4be0800000000000000ffffffffff07e1fc0800000000000000fffffffffe0dc3e82800000000000000fffffffffc1fc5f00080000000000000fffffffff83fefb00300000000000000fffffffff87ffff00140000000000000fffffffff0fffff00240040000000000ffffffffe1fffff004e0000000000000ffffffffd1fffff709c0000000000000ffffffffc3fffffe03a0000000000000ffffffff87fffffcd4a0008000000000ffffffffc7fff7ef9d40008000000000ffffffff0fffe7ffab90008000000000ffffffff8fffc7dfaa20020000000000ffffffffffffe3ffffffd7e063fd337dffffffffffffe1ffffffeff021ff307bffffffffffffc0ffffffcff030fd886effffffffffff807fffffc728107f1058ffffffffffff807fffffc208186e7070ffffffffffff007fffff810008521a24ffffffffffff003fffff800408341820fffffffffffe001ffffe00040c32a800fffffffffffe001ffffc000004360058fffffffffffe000fffbc000204193030fffffffffffc0007ffc8000006101880fffffffffffc0007ff80000002108860fffffffffff80007ff00000102000000fffffffffffc0003ff00000002000000fffffffffff80003fe00000003000800fffffffffff00001fe00000001000000fffffffffff00001fe00000081000000fffffffffff80001fc00000081000000fffffffffff00001fc00000081000000fffffffffff00001fc00000081800000fffffffffff00000fc000000c1800000fffffffffff80000fc000000c0800000fffffffffff000007c000000c0800000fffffffffff000007c00000040800000fffffffffff000007c000000c0800000ffffffffffe000007800000040800000ffffffffffe000007800000040800000fffffffffff8000070000000c0800000fffffffffff00000e0000000c0800000fffffffffff0000080000000c0800000ffffffffffe0000000000000e0000000ffffffffffe800000000000060800000fffffffffffc000000000000e3c001dffffffffffffc000000000000e380301ffffffffffffc000000000000e3c190fdfffffffffffc000000000000e3c08254fffffffffffc00006000000061c10038fffffffffff8000060000000e181005cfffffffffff8000060000000e3c00006fffffffffff80000f000000071800010fffffffffff80000f0000000f1804000fffffffffff80000f0000000e1800000fffffffffff80000f8000000e1800000fffffffffffc0000f8000000e1800000fffffffffffe000078000000e1000000fffffffffffe000078000281e1000000fffffffffffe000078000381e1000000fffffffffffe0000f8000741e1000000fffffffffffe0000fc0007f1c1000000fffffffffffe0001fc0007e1c0000000ffffffffffff0001fe000fe1c2000000ffffffffffff0003fe000fe1c2000000fffffffdffff0003ff0017e182000000fffffffdffff8007ff800fc386000000fffffffdffff8007ff800fc306000000fffffffeffff8007ff80078704000000fffffffeffffc00fff800fe704000000fffffffe7fffc01fffc00fce00000000fffffff67fffc01fffc01fce08000000fffffff77fffc01fffe01f8c00000000ffffffff3fffc03fffc01f9c10000000fffffffb9fffe07fffe03f3c00000000fffffffb8ffff07ffff03e3800000000fffffffdc7fff0fffff07e7800000000fffffffdcffffffffffffff040000000fffffffeeffffffffefffff040000000fffffffee7fffffffea7fee000000000ffffffff71fffffffcc3f5e080000000ffffffff71bfffffff07f0c100000000ffffffffb821ffffff87e18100000000ffffffff9c28ffffff87c18200000000ffffffffdc78ffffffcf830000000000ffffffffee7e47ffffcf860000000000ffffffffe71e07ffffff840000000000fffffffff78f03ffffff880000000000fffffffffb8f02ffffff100000000000fffffffffdc794ffffff200000000000fffffffffec3f83ffffe400000000000fffffffffe71fe0efffc800000000000ffffffffff38fe80fff2000000000000ffffffffff9c3fe09fec000000000000ffffffffffc71fe3bf90000000000000ffffffffffe30ffffe00000000000000fffffffffff803fff000000000000000fffffffffff800060000000000000000fffffffffff800000000000000000000fffffffffffc00000000000000000000fffffffffffe00000000000000000000ffffffffffffc0000000000000000000fffffffffffff0000000000000000000fffffffffffff80000000000000000007ffffffffffff80000000000000000006fffffffffffff0000000000000000006ffffffffffffff00000000000000000043ffffffffffff800000000000000000011ffffffffffd00000000000000000"`;

exports[`LeafSeg shall measure the leaf correctly 4`] = `"fbba4000001fffffffffffffffff0fff7fdf6000005ffbffffffffffffff0fff7eff6000037fffffffffffffffff0fff6fe3e00000ffffffffffffffffff0ffff2db800001ffffffffffffffffff0ffffefb400003ffffffffffffffffff0cfffed3600007ffffffffffffffffff0c7fffefa00025ffffffffffffffffff0cfffbeee00017ff07ffffffffffffff00fffae4200013f8030007ffffffffff0cffffe0a0000fc00000007fffffffff0fffffd3a00027000000001fffffffff0ffffd5220002c0000000007ffffffff0fffbf6ae000380000000001ffffffff0fffffd1e0006000000000007fffffff07fffca1a000c000000000003fffffff0fff7f0f60000000000000000fffffff0ffffd8c200000000810000007ffffff0fffff88400000000040000003ffffff0ffffea9c00000060000002801ffffff0cfff46ec00000080400007801ffffff007ff6f4c0000014040000f800ffffff0cffff9040000030000000f0007fffff0fffba014040006000000160003fffff0fff8000000000c000000120001fffff0fffec49400000c0000002a0001fffff0fffb001400001a000000280000fffff07fff80040000320000000400007ffff0ffff10040000200000000800007ffff0fffe20240000600000008900003ffff0fffe200c0000500000005000003ffff0fffc000c0000400000000400011ffff0477ffffdffc1fffffffffc0001dffff8060ffffdffc3fffffffff80001dffff8ec8ffffdff83fffffffff80000effff8fc8ffffdff87fedff7ffe00000effff8fccffffdff07fa9e13ffe00000fffff8fccffffdff07f01802ffe0000077fff8fccffffdfe0ff008007f2000003ffff87ccffffdfe0ff001004f0000003bfff8fccffffdfe1ff000000e8000007bfff8fccffffdfc1fe00000028000003ffff8fccffffdfc1fe00000010000003ffff8fccffffdf87fc00000000000001ffff8cccffffdf87f800000000000001ffff8048ffffdf87f800000000000005ffff8dc0ffffdf83f800000000000005ffff8fe0ffffdf87f000000000000007ffff8fffffffdf87f800000000000003ffff8fffffff9f87e800000000000007ffff8fffffff9f87e800000000000007ffff87ffffff9f07d800000000000007ffff8fffffff9f0ff800000000000003ffff8fffffff9f0f9000000000000003ffff87ffffff9f0fa000000000000003ffff87ffffff9f0f0000000000000003ffff805fffffbf0f0000000000000001ffff805fffffbf0f0000000000000001ffff847fffffbf0e0000000000000001ffff87ffffffbe2e0000000000000000ffff87ffffffbe2f0000000000000000ffff87ffffffbe3f00000000000000007fff87ffffff9e3f80000000000000007fff87ffffffbe3f80000000000000003fff87ffffffbf3ff0000000000000003fff87ffffffbf3ff0000000000000003fff87ffffffbf3ff8000000000000003fff87ffffffbf3ff8000000000000003fff807fffffbf3ff800000020000001bfff807fffffbf3ff800000068000009ffff8461ffffbf3ff80000007000000fffff86c0ffffbf3ff80000007000001fffff86c8ffffbf3ff80000007000001fffff87ccffffff3ff80000007000000fffff87ccffffff3ffc0000007000001fffff87ccffffff1ffc0000000000001fffff87ccffffff1ffe0000000000000fffff87ccffff7f1ffe0000000000000fffff87ccffff7f1ffe0000000000000fffff87ccffff7f9ffe0000000000000fffff844cffff7f8fff0000000000000fffff804cffff7f8fff0000004000000fffff8048ffff7f8fff0000008000000fffff8040ffff7f8fff000000e000000fffff80c0ffff7fc7ff00000040000015ffff8040ffff7fc7ff000000f0000009ffff80e7ffff7fc7ff800000f0000009ffff80c7ffff7fe3ff800001f8000001ffff80c3ffff7fe3ff800003f8000003ffff80c3ffff7fe3ffc00003fc000003bfff80c3ffff7ff1ffc0000ffc000023ffff8043ffff7ff1ffc0000ffc000027ffff8003ffff7ff0ffc0000ffc0000077fff8003ffff7ff8ffe0000ffc000007ffff8043fffffff87fe0000ffe00000effff8041fffffff87fe0000ffc00000effff8041fffffffc3ff0001ffe00001dffff8041fffffffc3ff8000ffe00001dffff8041fffffffe1ff8001ffe00002bffff8041fffffffe1ff8003fff00001bffff8041fffeffff0ffc003fff000017ffff8041fffeffff87fc007fff800027ffff8041fffeffff87fe007fff80002fffff8041fffeffffc3fe00ffffc0004fffff8001fffefffdc1fe00ffffc0009ff7ff8001fffeffffe1ff00ffffc0003fe3ff0001fffefffff0ff89ffffe0003f83ff0041fffeffff787fdfffffe0007f81ff0041fffeffff383ffffffff000ee03ff0041fffeffff1c0ffffffff001f205ff0041fffeffff6603fffffff003c400ff0040fffefffff300ffffffe003c006ff0040fffeffffe8803fffff800140001f0000fffeffffe0400fffff800000000f0000fffefffff02003fffe000000000f0001fffefffffc00007ff000000000070001fffefffffe0000000000000000060000fffeffffff8000000000000000060000fffeffffffc000000000000000020000fffeffffffd000000000000000000000fffefffffffd00000000000000000000fffcfffffff230000000000000000000fffcfffffffc50000000000000000000fffcfffffffb00000000000000000000fffcfffffffe00000000000000000000fffcffffffffe0a00000000000000000fffc7fffffffe3dc0000000000000000fffc3c7e087001800000000000000000"`;

exports[`LeafSeg shall retun a transparent image 1`] = `"fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffefffffffffffffffffffffffffffffffc7ffffffffffffffffffffffffffffffc3ffffffffffffffffffffffffffffffe9ffffffffffffffffffffffffffffffe8ffffffffffffffffffffffffffffffcc7fffffffffffffffffffffffffffffde7fffffffffffffffffffffffffffff9f3fffffffffffffffffffffffffffffbfbfffffffffffffffffffffffffffffbfbfffffffffffffffffffffffffffffbf9fffffffffffffffffffffffffffff3f9fffffffffffffffffffffffffffff7fcffffffffffffffffffffffffffffe7fcffffffffffffffffffffffffffffe7fcffffffffffffffffffffffffffffe7fcffffffffffffffffffffffffffffe7fdffffffffffffffffffffffffffffeffcffffffffffffffffffffffffffffeffcffffffffffffffffffffffffffffeffeffffffffffffffffffffffffffffeffe7fffffffffffffffffffffffffffeffe7fffffffffffffffffffffffffffcff97fffffffffffffffffffffffffffcff87fffffffffffffffffffffffffffcff9ffffffffffffffffffffffffffffeffcffffffffffffffffffffffffffffeffcffffffffffffffffffffffffffffeffeffffffffffffffffffffffffffffeffe7fffffffffffffffffffffffffffeffe7fffffffffffffffffffffffffffefff7fffffffffffffffffffffffffffefff7fffffffffffffffffffffffffffe9f87fffffffffffffffffffffffffffe1fbffffffffffffffffffffffffffffe1fbfffffffffffffffffffffffffffffdfbfffffffffffffffffffffffffffffdfbfffffffffffffffffffffffffffffdf3fffffffffffffffffffffffffffffcf3fffffffffffffffffffffffffffffe77fffffffffffffffffffffffffffffe77fffffffffffffffffffffffffffffe67ffffffffffffffffffffffffffffff6fffffffffffffffffffffffffffffff2fffffffffffffffffffffffffffffff0ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe7ffffffffffffffffffffffffffffffe1ffffffffffffffffffffffffffffffe4ffffffffffffffffffffffffffffffe6ffffffffffffffffffffffffffffffee7fffffffffffffffffffffffffffffce7fffffffffffffffffffffffffffffcf7fffffffffffffffffffffffffffff9f3fffffffffffffffffffffffffffff3f3fffffffffffffffffffffffffffff7f9ffffffffffffffffffffffffffffe7f9ffffffffffffffffffffffffffffeffdffffffffffffffffffffffffffffeff9ffffffffffffffffffffffffffffeff9ffffffffffffffffffffffffffffeff9ffffffffffffffffffffffffffffcff9ffffffffffffffffffffffffffffcffdffffffffffffffffffffffffffffcffdffffffffffffffffffffffffffffdffdffffffffffffffffffffffffffff9ffcffffffffffffffffffffffffffffbffcffffffffffffffffffffffffffffbffeffffffffffffffffffffffffffff87f6ffffffffffffffffffffffffffffe7f2ffffffffffffffffffffffffffffe7f0ffffffffffffffffffffffffffffeff3ffffffffffffffffffffffffffffeff3ffffffffffffffffffffffffffffeffbffffffffffffffffffffffffffffeffbffffffffffffffffffffffffffffcffbffffffffffffffffffffffffffffcffbffffffffffffffffffffffffffffc3fbffffffffffffffffffffffffffffe3e3fffffffffffffffffffffffffffff9e3fffffffffffffffffffffffffffff9cffffffffffffffffffffffffffffff9cffffffffffffffffffffffffffffffddffffffffffffffffffffffffffffffddffffffffffffffffffffffffffffffd9ffffffffffffffffffffffffffffffc9ffffffffffffffffffffffffffffffcbffffffffffffffffffffffffffffffe3ffffffffffffffffffffffffffffffe3fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"`;
